src/dns/named.rb

```diff
@@ -0,0 +1,141 @@
+require 'templater'
+require 'exceptions'
+require 'metal_log'
+
+module Metalware
+  module DNS
+    class Named
+      def initialize(metalware_config)
+        @config = metalware_config
+      end
+
+      def update
+        setup unless setup?
+        render_repo_named_conf
+        each_network do |zone, net|
+          render_zone_template('forward', zone, net)
+          render_zone_template('reverse', zone, net)
+        end
+      end
+
+      private
+
+      attr_reader :config
+
+      def file_path
+        @file_path ||= FilePath.new(config)
+      end
+
+      def setup
+        check_external_dns_is_set
+        render_base_named_conf
+        restart_named
+      end
+
+      def setup?
+        exit_code = SystemCommand.run_raw('systemctl status named')[:status]
+        MetalLog.info "systemctl status named, exit code: #{exit_code}"
+        exit_code.success?
+      end
+
+      EXTERNAL_DNS_MSG = <<~EOF.strip_heredoc
+        Cannot setup 'named' DNS type as 'externaldns' has not been set in the
+        'domain.yaml' repo config file.
+      EOF
+
+      def check_external_dns_is_set
+        raise MissingExternalDNS, EXTERNAL_DNS_MSG unless repo_config[:externaldns]
+      end
+
+      def repo_config
+        @repo_config ||= Templating::RepoConfigParser
+                         .parse_for_domain(config: config,
+                                           include_groups: false)
+                         .inspect
+      end
+
+      def render_base_named_conf
+        Templater.render_to_file(config, file_path.named_template, file_path.base_named)
+      end
+
+      def render_repo_named_conf
+        FileUtils.mkdir_p(File.dirname(file_path.metalware_named))
+        Templater.render_to_file(config,
+                                 file_path.template_path('named'),
+                                 file_path.metalware_named)
+      end
+
+      # TODO: These commands will break hosts DNS. Might be a good idea to run
+      # similar commands for hosts
+      RESTART_NAMED_CMDS = <<~EOF.strip_heredoc
+        systemctl disable dnsmasq
+        systemctl stop dnsmasq
+        systemctl enable named
+        systemctl restart named
+      EOF
+
+      # DO NOT RENAME, DANGEROUS FOR TESTS
+      def restart_named
+        MetalLog.info 'Restarting named'
+        SystemCommand.run(RESTART_NAMED_CMDS)
+      end
+
+      def each_network
+        repo_config[:networks]&.each { |data| yield data }
+      end
+
+      def named_zone_hash(zone, net)
+        {
+          alces_named: {
+            zone: zone,
+            net: net,
+          },
+        }
+      end
+
+      def render_zone_template(direction, zone, net)
+        zone_name = case direction
+                    when 'forward'
+                      net[:named_fwd_zone]
+                    when 'reverse'
+                      net[:named_rev_zone]
+                    end
+        zone_hash = named_zone_hash(zone, net)
+        return unless zone_name
```

Comment:
Don't think we need this line? The tests pass without it, and afaics it shouldn't be reached. If we do reach it by accident in future though, it would be better to fail noisily because this is `nil` rather than silently I think.